File: app/src/main/java/com/hustunique/morii/content/ContentActivity.java
package com.hustunique.morii.content;

import static com.hustunique.morii.util.MyApplication.musicDiaryList;
import static com.hustunique.morii.util.MyApplication.musicTabList;
import static com.hustunique.morii.util.MyApplication.soundItemList;

import android.content.Intent;
import android.os.Bundle;
import android.transition.AutoTransition;
import android.transition.Explode;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.SeekBar;
import android.widget.TextView;

import androidx.cardview.widget.CardView;

import com.bumptech.glide.Glide;
import com.hustunique.morii.database.DiaryInfo;
import com.hustunique.morii.database.SoundItemInfo;
import com.hustunique.morii.home.MainActivity;
import com.hustunique.morii.home.MusicDiaryItem;
import com.hustunique.morii.util.AudioExoPlayerUtil;
import com.hustunique.morii.util.BaseActivity;
import com.hustunique.morii.util.DatabaseUtil;
import com.hustunique.morii.util.onReadyListener;

import java.util.List;

import morii.R;

public class ContentActivity extends BaseActivity {

    private CardView deleteButton;
    private CardView finishButton;
    private CardView pauseMusic;
    private CardView goBack;
    private TextView titleBarText;
    private MusicDiaryItem musicDiaryItem;
    private static final String TAG = "ContentActivity";
    private Intent intent;
    int newItem = 0;
    private ImageView imageView2;
    private TextView startTime;
    private ProgressBar progressBar;
    private SeekBar seekBar;
    private int Position = (int) AudioExoPlayerUtil.getCurrentPosition();
    private TextView title;
    private TextView article;
    private TextView date;
    private TextView tag;
    private TextView endTime;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().requestFeature(Window.FEATURE_ACTIVITY_TRANSITIONS);
        getWindow().setSharedElementExitTransition(new AutoTransition());
        getWindow().setEnterTransition(new Explode());
        setContentView(R.layout.activity_content);
        Log.d(TAG, "onCreate: ");
        intent = getIntent();
        newItem = intent.getIntExtra("NewItem", 0);
        viewBinding();
        initMusicDiaryContent();
        setCallbacks();
    }

    private void viewBinding() {
        pauseMusic = findViewById(R.id.MusicPlay);
        imageView2 = findViewById(R.id.music_pause);
        startTime = findViewById(R.id.StartTime);
        progressBar = findViewById(R.id.MusicLine);
        seekBar = findViewById(R.id.SeekBar);
        titleBarText = findViewById(R.id.title_content);
        finishButton = findViewById(R.id.finishButton);
        deleteButton = findViewById(R.id.deleteButton);
        goBack = findViewById(R.id.backLayout_content);
        title = findViewById(R.id.musicDiaryTitle);
        article = findViewById(R.id.diaryContent);
        date = findViewById(R.id.musicDiaryDate);
        tag = findViewById(R.id.musicDiaryTag);
        endTime = findViewById(R.id.EndTime);
        progressBar = findViewById(R.id.MusicLine);
        seekBar = findViewById(R.id.SeekBar);
    }

    private void setCallbacks() {
        if (AudioExoPlayerUtil.isPlaying() || newItem == 0) {
            Glide.with(this).load(R.drawable.outline_pause_24).into(imageView2);
            Log.d(TAG, "Playing");
        } else {
            Log.d(TAG, "not Playing");
            Glide.with(this).load(R.drawable.round_play_arrow_24).into(imageView2);
        }
        pauseMusic.setOnClickListener(view -> {
            if (AudioExoPlayerUtil.isPlaying()) {
                AudioExoPlayerUtil.pauseAllPlayers();
                Glide.with(this).load(R.drawable.round_play_arrow_24).into(imageView2);
            } else {
                AudioExoPlayerUtil.startAllPlayers();
                play();
                Glide.with(this).load(R.drawable.outline_pause_24).into(imageView2);

            }
        });
        if (newItem == 0) {
            finishButton.setVisibility(View.GONE);
            deleteButton.setVisibility(View.GONE);
        } else {
            titleBarText.setText("预览");
            finishButton.setOnClickListener(v -> {
                newItem = 0;
                createMusicDiary();
                Intent backIntent = new Intent(ContentActivity.this, MainActivity.class);
                startActivity(backIntent);
            });
            deleteButton.setOnClickListener(v -> {
                newItem = 0;
                Intent backIntent = new Intent(ContentActivity.this, MainActivity.class);
                startActivity(backIntent);
            });
        }
        goBack.setOnClickListener(view -> onBackPressed());
        if (newItem == 1) {
            initProgressBar(AudioExoPlayerUtil.getDuration());
        } else {
            AudioExoPlayerUtil.setListener(new onReadyListener() {
                @Override
                public void onReady(long duration) {
                    super.onReady(duration);
                    Log.d(TAG, "onReady: " + duration);
                    initProgressBar(duration);
                }
            });
        }
        seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                String sss, mmm;
                sss = String.valueOf(progress / 1000 % 60);
                mmm = String.valueOf(progress / 60 / 1000);
                if (sss.length() < 2) sss = "0" + sss;
                if (mmm.length() < 2) mmm = "0" + mmm;
                startTime.setText(mmm + ":" + sss);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });
    }

    private void initMusicDiaryContent() {
        StringBuilder builder = new StringBuilder();
        musicDiaryItem = (MusicDiaryItem) intent.getSerializableExtra("diary");
        builder.append(musicTabList.get(musicDiaryItem.getMusicTabId()).getEmotion()).append(" ");
        if (newItem == 0) {
            Log.d(TAG, "initMusicDiaryContent: newItem=0");
            AudioExoPlayerUtil.playMusic(musicDiaryItem.getMusicTabId());
            List<SoundItemInfo> list = musicDiaryItem.getSoundItemInfoList();
            for (SoundItemInfo info : list) {
                Log.d(TAG, info.soundItemId + " position:" + info.soundItemPosition);
                AudioExoPlayerUtil.setSoundPlayer(info.soundItemId, info.soundItemPosition);
                AudioExoPlayerUtil.startSoundPlayer(info.soundItemPosition);
                builder.append(soundItemList.get(info.soundItemId).getSoundName()).append(" ");
            }
        } else {
            for (SoundItemInfo info : musicDiaryItem.getSoundItemInfoList()) {
                builder.append(soundItemList.get(info.soundItemId).getSoundName()).append(" ");
            }
            initProgressBar(AudioExoPlayerUtil.getDuration());
        }
        title.setText(musicDiaryItem.getTitle());
        article.setText(musicDiaryItem.getArticle());
        date.setText(musicDiaryItem.getDate());
        tag.setText(builder.toString());
        // get the information
        ImageView imageView = findViewById(R.id.PhotoShow);
        String imagePath = musicDiaryItem.getImagePath();
        Log.d(TAG, "onCreate: " + imagePath);

        if (imagePath != null) {
            Glide.with(this).load(imagePath).into(imageView);
        } else {
            Glide.with(this).load(musicTabList.get(musicDiaryItem.getMusicTabId())
                    .getImageResId()).into(imageView);
        }
    }

    private void initProgressBar(long Duration) {
        progressBar.setMax((int) Duration);
        seekBar.setMax((int) Duration);
        String sss, mmm;
        sss = String.valueOf(Duration / 1000 % 60);
        mmm = String.valueOf(Duration / 60 / 1000);
        if (sss.length() < 2) sss = "0" + sss;
        if (mmm.length() < 2) mmm = "0" + mmm;
        endTime.setText(mmm + ":" + sss);
        play();
    }


    private void createMusicDiary() {
        DiaryInfo diaryInfo = new DiaryInfo(musicDiaryItem);
        long diaryInfoId = DatabaseUtil.insertDiaryInfo(diaryInfo);
        musicDiaryItem.setItemID(diaryInfoId);
        for (SoundItemInfo info : musicDiaryItem.getSoundItemInfoList()) {
            info.diaryInfoId = diaryInfoId;
            DatabaseUtil.insertSoundItemInfo(info);
        }
        musicDiaryList.add(musicDiaryItem);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (newItem == 0) {
            AudioExoPlayerUtil.pauseAllPlayers();
            AudioExoPlayerUtil.resetAllSoundPlayers();
        }
    }

    private void play() {

        Thread thread = new Thread(new goThread());
        thread.start();
    }

    class goThread implements Runnable {
        boolean isPlaying = true;

        @Override
        public void run() {
            while (isPlaying) {
                try {
                    runOnUiThread(() -> {
                        isPlaying = AudioExoPlayerUtil.isPlaying();
                        if (isPlaying)
                            Position = (int) AudioExoPlayerUtil.getCurrentPosition();
                    });
                    seekBar.setProgress(Position);
                    progressBar.setProgress(Position);
                    Thread.sleep(100);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }

            }

        }
    }
}

File: app/src/main/java/com/hustunique/morii/content/ContentActivity.kt
package com.hustunique.morii.content

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.transition.AutoTransition
import android.transition.Explode
import android.util.Log
import android.view.*
import android.widget.*
import android.widget.SeekBar.OnSeekBarChangeListener
import androidx.core.content.FileProvider
import com.bumptech.glide.Glide
import com.hustunique.morii.database.DiaryInfo
import com.hustunique.morii.home.MainActivity
import com.hustunique.morii.home.MusicDiaryItem
import com.hustunique.morii.util.*
import com.hustunique.morii.util.MyApplication.Companion.musicTabList
import morii.R
import morii.databinding.ActivityContentBinding
import java.io.File


class ContentActivity : BaseActivity() {
    private lateinit var binding: ActivityContentBinding
    private lateinit var musicDiaryItem: MusicDiaryItem
    private lateinit var handler: Handler
    private var newItem = 0
    private var position = AudioExoPlayerUtil.currentPosition.toInt()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.requestFeature(Window.FEATURE_ACTIVITY_TRANSITIONS)
        window.sharedElementExitTransition = AutoTransition()
        window.enterTransition = Explode()
        binding = ActivityContentBinding.inflate(layoutInflater)
        setContentView(binding.root)
        newItem = intent.getIntExtra("NewItem", 0)
        initUI()
        initMusicDiaryContent()
        setCallbacks()
    }

    private fun initUI() {
        if (AudioExoPlayerUtil.isPlaying || newItem == 0) {
            Glide.with(this).load(R.drawable.outline_pause_24)
                .into(binding.progressbarContent.image)
        } else {
            Glide.with(this).load(R.drawable.round_play_arrow_24)
                .into(binding.progressbarContent.image)
        }
        if (newItem == 0) {
            binding.finishButton.visibility = View.GONE
            binding.deleteButton.visibility = View.GONE
            binding.completeLayoutContent.visibility = View.VISIBLE
        } else {
            binding.titleContent.text = "预览"
            Glide.with(this).load(R.drawable.ic_icon_left_design).into(binding.backButtonImage)
            binding.backButtonText.text = "上一步"
        }
    }

    private fun setCallbacks() {

        binding.progressbarContent.button
            .setOnClickListener {
                if (AudioExoPlayerUtil.isPlaying) {
                    AudioExoPlayerUtil.pauseAllPlayers()
                    Glide.with(this).load(R.drawable.round_play_arrow_24)
                        .into(binding.progressbarContent.image)
                } else {
                    AudioExoPlayerUtil.startAllPlayers()
                    play()
                    Glide.with(this).load(R.drawable.outline_pause_24)
                        .into(binding.progressbarContent.image)
                }
            }

        if (newItem == 1) {
            binding.finishButton.setOnClickListener {
                createMusicDiary()
                backToMainActivity()
            }
            binding.deleteButton.setOnClickListener { v: View? ->
                backToMainActivity()
            }
            initProgressBar(AudioExoPlayerUtil.getDuration())

        } else {
            AudioExoPlayerUtil.setListener(object : OnReadyListener() {
                override fun onReady(duration: Long) {
                    super.onReady(duration)
                    Log.d(TAG, "onReady: $duration")
                    initProgressBar(duration)
                }
            })
        }
        binding.backLayoutContent.setOnClickListener { onBackPressed() }
        binding.completeLayoutContent.setOnClickListener {
            Toast.makeText(this, "正在生成音频文件", Toast.LENGTH_SHORT).show()
            AudioProcessor.makeAudioMix(musicDiaryItem, handler)
        }

        binding.progressbarContent.SeekBar.setOnSeekBarChangeListener(object :
            OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                var sss: String
                var mmm: String
                sss = (progress / 1000 % 60).toString()
                mmm = (progress / 60 / 1000).toString()
                if (sss.length < 2) sss = "0$sss"
                if (mmm.length < 2) mmm = "0$mmm"
                binding.progressbarContent.StartTime.text = "$mmm:$sss"
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })
        handler = Handler(this.mainLooper) {
            val uri = FileProvider.getUriForFile(
                this,
                "com.hustunique.morii.provider",
                File(it.obj as String)
            )
            shareAudioFile(uri)
            return@Handler true
        }
    }

    private fun initMusicDiaryContent() {
        val builder = StringBuilder()
        musicDiaryItem = intent.getSerializableExtra("diary") as MusicDiaryItem
        builder.append(
            musicTabList[musicDiaryItem.musicTabId].emotion
        ).append(" ")

        if (newItem == 0) {
            Log.d(TAG, "initMusicDiaryContent: newItem=0")
            AudioExoPlayerUtil.playMusic(musicDiaryItem.musicTabId)
            for (info in musicDiaryItem.soundItemInfoList) {
                Log.d(TAG, info.soundItemId.toString() + " position:" + info.soundItemPosition)
                AudioExoPlayerUtil.setSoundPlayer(info.soundItemId, info.soundItemPosition)
                AudioExoPlayerUtil.startSoundPlayer(info.soundItemPosition)
                builder.append(
                    MyApplication.soundItemList[info.soundItemId].soundName
                ).append(" ")
            }
        } else {
            for (info in musicDiaryItem.soundItemInfoList) {
                builder.append(
                    MyApplication.soundItemList[info.soundItemId].soundName
                ).append(" ")
            }
            initProgressBar(AudioExoPlayerUtil.getDuration())
        }
        binding.musicDiaryTitle.text = musicDiaryItem.title
        binding.diaryContent.text = musicDiaryItem.article
        binding.musicDiaryDate.text = musicDiaryItem.date
        binding.musicDiaryTag.text = builder.toString()
        if (musicDiaryItem.imagePath != null)
            Glide.with(this).load(musicDiaryItem.imagePath).into(binding.PhotoShow)
        else
            Glide.with(this).load(musicTabList[musicDiaryItem.musicTabId].imageResId)
                .into(binding.PhotoShow)
    }

    private fun initProgressBar(Duration: Long) {
        binding.progressbarContent.MusicLine.max = Duration.toInt()
        binding.progressbarContent.SeekBar.max = Duration.toInt()
        var sss: String
        var mmm: String
        sss = (Duration / 1000 % 60).toString()
        mmm = (Duration / 60 / 1000).toString()
        if (sss.length < 2) sss = "0$sss"
        if (mmm.length < 2) mmm = "0$mmm"
        binding.progressbarContent.EndTime.text = "$mmm:$sss"
        play()
    }

    private fun createMusicDiary() {
        val diaryInfo = DiaryInfo(musicDiaryItem)
        val diaryInfoId = DatabaseUtil.insertDiaryInfo(diaryInfo)
        musicDiaryItem.itemID = diaryInfoId
        for (info in musicDiaryItem.soundItemInfoList) {
            info.diaryInfoId = diaryInfoId
            DatabaseUtil.insertSoundItemInfo(info)
        }
        MyApplication.musicDiaryList.add(musicDiaryItem)
    }

    private fun shareAudioFile(uri: Uri) {
        val shareIntent: Intent = Intent().apply {
            action = Intent.ACTION_SEND
            putExtra(Intent.EXTRA_STREAM, uri)
            type = "*/*"
        }
        startActivity(Intent.createChooser(shareIntent, "分享音频文件"))

    }

    private fun play() {
        val thread = Thread(goThread())
        thread.start()
    }

    internal inner class goThread : Runnable {
        var isPlaying = true
        override fun run() {
            while (isPlaying) {
                try {
                    runOnUiThread {
                        isPlaying = AudioExoPlayerUtil.isPlaying
                        if (isPlaying) position = AudioExoPlayerUtil.currentPosition.toInt()
                    }
                    binding.progressbarContent.SeekBar.progress = position
                    binding.progressbarContent.MusicLine.progress = position
                    Thread.sleep(100)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
            }
        }
    }

    override fun onDestroy() {
        if (newItem == 0) {
            AudioExoPlayerUtil.pauseAllPlayers()
            AudioExoPlayerUtil.resetAllSoundPlayers()
        }
        super.onDestroy()
    }

    private fun backToMainActivity() {
        val backIntent = Intent(this@ContentActivity, MainActivity::class.java)
        startActivity(backIntent)
        AudioExoPlayerUtil.pauseAllPlayers()
        AudioExoPlayerUtil.resetAllSoundPlayers()
    }

    companion object {
        private const val TAG = "ContentActivity"
    }

}